src/diffusers/quantizers/bitsandbytes/bnb_quantizer.py

```diff
@@ -0,0 +1,579 @@
+        param_name: str,
+        state_dict: Dict[str, Any],
+        **kwargs,
+    ) -> bool:
+        import bitsandbytes as bnb
+
+        module, tensor_name = get_module_from_name(model, param_name)
+        if isinstance(module._parameters.get(tensor_name, None), bnb.nn.Params4bit):
+            # Add here check for loaded components' dtypes once serialization is implemented
+            return True
+        elif isinstance(module, bnb.nn.Linear4bit) and tensor_name == "bias":
+            # bias could be loaded by regular set_module_tensor_to_device() from accelerate,
+            # but it would wrongly use uninitialized weight there.
+            return True
+        else:
+            return False
+
+    def create_quantized_param(
+        self,
+        model: "ModelMixin",
+        param_value: "torch.Tensor",
+        param_name: str,
+        target_device: "torch.device",
+        state_dict: Dict[str, Any],
+        unexpected_keys: Optional[List[str]] = None,
+    ):
+        import bitsandbytes as bnb
+
+        module, tensor_name = get_module_from_name(model, param_name)
+
+        if tensor_name not in module._parameters:
+            raise ValueError(f"{module} does not have a parameter or a buffer named {tensor_name}.")
+
+        old_value = getattr(module, tensor_name)
+
+        if tensor_name == "bias":
+            if param_value is None:
+                new_value = old_value.to(target_device)
+            else:
+                new_value = param_value.to(target_device)
+
+            new_value = torch.nn.Parameter(new_value, requires_grad=old_value.requires_grad)
+            module._parameters[tensor_name] = new_value
+            return
+
+        if not isinstance(module._parameters[tensor_name], bnb.nn.Params4bit):
+            raise ValueError("this function only loads `Linear4bit components`")
+        if (
+            old_value.device == torch.device("meta")
+            and target_device not in ["meta", torch.device("meta")]
+            and param_value is None
+        ):
+            raise ValueError(f"{tensor_name} is on the meta device, we need a `value` to put in on {target_device}.")
+
+        # construct `new_value` for the module._parameters[tensor_name]:
+        if self.pre_quantized:
+            # 4bit loading. Collecting components for restoring quantized weight
+            # This can be expanded to make a universal call for any quantized weight loading
+
+            if not self.is_serializable:
+                raise ValueError(
+                    "Detected int4 weights but the version of bitsandbytes is not compatible with int4 serialization. "
+                    "Make sure to download the latest `bitsandbytes` version. `pip install --upgrade bitsandbytes`."
+                )
+
+            if (param_name + ".quant_state.bitsandbytes__fp4" not in state_dict) and (
+                param_name + ".quant_state.bitsandbytes__nf4" not in state_dict
+            ):
```

Comment:
@SunMarc 

Do we have to additionally pop these keys out? Because [when a quantized checkpoint](https://huggingface.co/sayakpaul/flux.1-dev-nf4-with-bnb-integration) is loaded, it complains about 

```bash
Some weights of the model checkpoint were not used when initializing FluxTransformer2DModel: 
 ['context_embedder.weight.absmax, context_embedder.weight.quant_map, context_embedder.weight.quant_state.bitsandbytes__nf4, norm_out.linear.weight.absmax, norm_out.linear.weight.quant_map, norm_out.linear.weight.quant_state.bitsandbytes__nf4, proj_out.weight.absmax, proj_out.weight.quant_map, proj
...
```

Reply:
I know `transformers` handles this sophisticatedly in:
https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_utils.py

I imagine we could also do something similar here:
https://github.com/huggingface/diffusers/blob/d098d073e0b67247ecc7e9c98e6614346dc39008/src/diffusers/models/model_loading_utils.py#L184

But I don't have a clear idea on how we could leverage `hf_quantizer` and filter out stats related keys in a clean manner. Guidance would be appreciated.